Agnos doesn't build with newer versions of Boost

Agnos uses methods in the details namespace that have been moved in newer versions of boost:

```
g++ -o src\protocol.o -c src\protocol.cpp
src\protocol.cpp: In member function 'int32_t agnos::protocol::ClientUtils::get_seq()':
src\protocol.cpp:281:32: error: 'boost::interprocess::detail' has not been declared
scons: *** [src\protocol.o] Error 1
scons: building terminated because of errors.
```

--- a/libagnos/cpp/src/protocol.cpp
+++ b/libagnos/cpp/src/protocol.cpp
@@ -20,7 +20,15 @@
 
 #include "protocol.hpp"
 #include <boost/interprocess/detail/atomic.hpp>
+#include <boost/detail/workaround.hpp>
 
+#include <boost/version.hpp>
+
+#if BOOST_WORKAROUND(BOOST_VERSION, >= 104800)
+namespace bid = boost::interprocess::ipcdetail;
+#else
+namespace bid = boost::interprocess::detail;
+#endif
 
 namespace agnos
 {
@@ -278,7 +286,7 @@ namespace agnos
 
 		int32_t ClientUtils::get_seq()
 		{
-			return boost::interprocess::detail::atomic_inc32((boost::uint32_t*)&_seq);
+			return bid::atomic_inc32((boost::uint32_t*)&_seq);
 		}
 
 		shared_ptr<PackedException> ClientUtils::load_packed_exception()

--- a/libagnos/cpp/src/servers.cpp
+++ b/libagnos/cpp/src/servers.cpp
@@ -107,7 +107,7 @@ namespace agnos
 
 		void LibraryModeServer::serve()
 		{
-			shared_ptr<SocketTransportFactory> factory = boost::shared_static_cast<SocketTransportFactory>(transport_factory);
+			shared_ptr<SocketTransportFactory> factory = boost::dynamic_pointer_cast<SocketTransportFactory>(transport_factory);
 			std::cout << "AGNOS" << std::endl;
 			std::cout << factory->acceptor->local_endpoint().address().to_string() << std::endl;
 			std::cout << factory->acceptor->local_endpoint().port() << std::endl;

